### PR TITLE
cudnn: 7.6.5 -> 8.3.2

### PR DESCRIPTION
**Pull request by samuela**
###### Description of changes
Note that this PR is targeting staging instead of master because the PyTorch version on staging (v1.11.0) is more amenable to cudnn upgrades than the PyTorch version on master (v1.10.2).

This PR also addresses https://github.com/NixOS/nixpkgs/pull/166359#issuecomment-1082737458 to follow up on my sworn promise to rid nixpkgs of cudnn_8_1_cudatoolkit_11.

Running tools like nixpkgs-review on this PR is going to be tricky atm due to https://github.com/NixOS/nixpkgs/issues/166444, which thus far we are led to believe is a temporary CDN interruption. However, I have built pytorchWithCuda successfully since I already had cudnn builds cached. 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/science/math/cudnn/default.nix [hunk 1]
```diff
@@ -81,31 +81,27 @@ rec {
   cudnn_8_1_cudatoolkit_11_2 = cudnn_8_1_cudatoolkit_10_2.override { cudatoolkit = cudatoolkit_11_2; };
 
   cudnn_8_1_cudatoolkit_10 = cudnn_8_1_cudatoolkit_10_2.override { cudatoolkit = cudatoolkit_10; };
-  cudnn_8_1_cudatoolkit_11 = cudnn_8_1_cudatoolkit_10_2.override { cudatoolkit = cudatoolkit_11; };
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/science/math/cudnn/default.nix:84]: Deleted from `cudnn/default.nix` but, apparently, still `inherit`-ed in `all-packages.nix`: https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/101

**Review by SomeoneSerge — CHANGES_REQUESTED** [review]

**Comment by samuela** [comment]
Notes:
* I uncovered an issue with cupy where it would pull in both cudatoolkit_11_5 and cudatoolkit_11_4. This was bc `cutensor_cudatoolkit_11 = cutensor_cudatoolkit_11_4;` instead of overriding with cudatoolkit_11. I fixed this in https://github.com/NixOS/nixpkgs/pull/166533/commits/d7f9d3a6db3968b26a240ca3b6025663eb1cf7db. To avoid this failure in the future I added asserts in cupy/default.nix to double check that we're using consistent cudatoolkit versions across all of our dependencies.
* cudatoolkit_11 was pointing to cudatoolkit_11_6 due to https://github.com/NixOS/nixpkgs/pull/165021. This has already been fixed on master (https://github.com/NixOS/nixpkgs/pull/166359), but not on staging. I fixed inhttps://github.com/NixOS/nixpkgs/pull/166533/commits/259ac51d2f66425880ff094822dee6a01f4acab1 and added a comment noting the dangers of updating cudatoolkit_11.
* cutensor is currently at 1.3.1, but the latest version is 1.5.0. Refraining from updating in this PR.
* cupy is currently at 10.2.0, but the latest version is 10.3.0. Refraining from updating in this PR.

File: pkgs/development/libraries/science/math/cudnn/default.nix (hunk 1 above)
Comment (samuela) [reply pkgs/development/libraries/science/math/cudnn/default.nix:84]: Ah good catch! fixed in https://github.com/NixOS/nixpkgs/pull/166533/commits/5e686a3a78df40f06b0c1dc064b130fe4a9d9318

**Comment by samuela** [comment]
I'm going to close this in favor of https://github.com/NixOS/nixpkgs/pull/166784 which is (roughly) the same thing but targeting master instead of staging. If that fails for some reason, we can revive this PR.

**Pull request closed without merging**